include/deal.II/lac/solver_gmres.h

```diff
@@ -806,208 +946,349 @@ namespace internal
+              if (delayed_reorthogonalization)
+                {
+                  h(dim + dim - 1) += last_vector * last_vector;
+                  h(dim + dim) += vvec * vvec;
+                }
+              for (unsigned int i = 0; i < dim - 1; ++i)
+                {
+                  const double temp =
+                    block(orthogonal_vectors[i], b).local_element(j);
+                  h(i) += temp * vvec;
+                  if (delayed_reorthogonalization)
+                    h(dim + i) += temp * last_vector;
+                }
+            }
         }
 
       Utilities::MPI::sum(h, block(vv, 0).get_mpi_communicator(), h);
     }
 
 
 
-    template <typename VectorType,
+    template <bool delayed_reorthogonalization,
+              typename VectorType,
               std::enable_if_t<
                 !is_dealii_compatible_distributed_vector<VectorType>::value,
                 VectorType> * = nullptr>
     double
-    subtract_and_norm(
-      const unsigned int dim,
-      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
-                           &orthogonal_vectors,
-      const Vector<double> &h,
-      VectorType           &vv)
+    subtract_and_norm(const unsigned int            dim,
+                      const TmpVectors<VectorType> &orthogonal_vectors,
+                      const Vector<double>         &h,
+                      VectorType                   &vv)
     {
       Assert(dim > 0, ExcInternalError());
 
+      VectorType &last_vector =
+        const_cast<VectorType &>(orthogonal_vectors[dim - 1]);
       for (unsigned int i = 0; i < dim - 1; ++i)
-        vv.add(-h(i), orthogonal_vectors[i]);
+        {
+          if (delayed_reorthogonalization && i + 2 < dim)
+            last_vector.add(-h(dim + i), orthogonal_vectors[i]);
+          vv.add(-h(i), orthogonal_vectors[i]);
+        }
 
-      return std::sqrt(
-        vv.add_and_dot(-h(dim - 1), orthogonal_vectors[dim - 1], vv));
+      if (delayed_reorthogonalization)
+        {
+          if (dim > 1)
+            last_vector.sadd(1. / h(dim + dim - 1),
+                             -h(dim + dim - 2) / h(dim + dim - 1),
+                             orthogonal_vectors[dim - 2]);
+
+          // h(dim + dim) is lucky breakdown
+          const double scaling_factor_vv =
+            h(dim + dim) > 0.0 ? 1. / (h(dim + dim - 1) * h(dim + dim)) :
+                                 1. / (h(dim + dim - 1) * h(dim + dim - 1));
+          vv.sadd(scaling_factor_vv,
+                  -h(dim - 1) * scaling_factor_vv,
+                  last_vector);
+
+          // the delayed reorthogonalization computes the norm from other
+          // quantities
+          return std::numeric_limits<double>::signaling_NaN();
+        }
+      else
+        return std::sqrt(
+          vv.add_and_dot(-h(dim - 1), orthogonal_vectors[dim - 1], vv));
     }
 
 
 
-    template <typename VectorType,
+    template <bool delayed_reorthogonalization,
```

Comment:
How much slower is the code without the template argument?

Reply:
I use fixed-size arrays that depend on this value to use as good register loop blocking reasonably possibly, as I want to load 12 values in the classical Gram-Schmidt and 6 values in the new delayed orthogonalization code path (due to 2 parallel accumulation, it gets 12 registers for partial results or input again). I can try with 6 in both cases if you want and measure for a case from cache and a case from main memory (the latter should be memory bandwidth bound, but the shorter the loop, the more likely are visible effects from the translation lookaside buffer or prefetching limitations). I guess this just made a case for measuring, didn't it?

Reply:
Thank you for the explanation. I am fine with the current status.

Reply:
I ran a short benchmark, using a locally owned size of 1000 (cached) or 1000000 (not cached). Here is the result with the current code on 72 Intel Ice Lake cores:
```
Timing MGS size 1000: 5.63975e-05 5.63983e-05 s or 326.823 GB/s
Timing MGS size 1000000: 0.11909 0.119101 s or 154.773 GB/s
Timing CGS size 1000: 1.47914e-05 1.47922e-05 s or 1246.13 GB/s
Timing CGS size 1000000: 0.0689767 0.0689878 s or 267.221 GB/s
Timing DCGS2 size 1000: 1.22783e-05 1.2279e-05 s or 1501.19 GB/s
Timing DCGS2 size 1000000: 0.066445 0.0664467 s or 277.402 GB/s
```
Here CGS is classical Gram-Schmidt, MGS is modified G-S, and DCGS2 is the classical Gram-Schmidt method with delayed reorthogonalization (proposed patch). On this system, the new algorithm runs even faster than the classical Gram-Schmidt approach, whereas I observed the opposite on my notebook. Without the template, I see 
```
Timing MGS size 1000: 5.56948e-05 5.56949e-05 s or 330.947 GB/s
Timing MGS size 1000000: 0.121417 0.121438 s or 151.808 GB/s
Timing CGS size 1000: 1.5118e-05 1.51188e-05 s or 1219.21 GB/s
Timing CGS size 1000000: 0.0689767 0.0689878 s or 267.221 GB/s
Timing DCGS2 size 1000: 1.24458e-05 1.24465e-05 s or 1480.98 GB/s
Timing DCGS2 size 1000000: 0.0647302 0.0647319 s or 284.751 GB/s
```
As you see, the difference is not big (by comparing to MGS, we can see the effect of noise). From L2 cache, the performance is slightly lower, but not spectacularly so. I think we can keep the templates for now, we can always adjust later, I had to change 15 lines.

Reply:
Thanks for the investigation!